### PR TITLE
rename Element.visitChildrenForSemantics to visitOnstageChildren; fix DropdownButton

### PR DESCRIPTION
**Pull request by yjbanov**
- rename `Element.visitChildrenForSemantics` to `visitOnstageChildren` and
  repurpose it for identifying interactive widgets
- mark hidden parts of `DropdownButton` as offstage

/cc @goderbauer @Hixie

### CODE REVIEW

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -2595,10 +2595,22 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
...
+  /// to interact with it. The precise definition of "interaction" is specific
+  /// to the widget, and may include hit testing, semantics, visibility as well
+  /// as other properties.
+  ///
+  /// The default implementation defers to [visitChildren] and therefore treats
+  /// the element as onstage.
+  ///
+  /// See also:
+  ///
+  /// - [Offstage]
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart]: nit: I often find it helpful to have a sentence fragment here explaining why I should also look at Offstage and visitChildrenForSemantics.

**Review by goderbauer — APPROVED** [review]

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (yjbanov) [inline packages/flutter/lib/src/widgets/framework.dart]: Done.

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 2]
```diff
@@ -2566,16 +2566,19 @@ class RenderIgnorePointer extends RenderProxyBox {
...
-/// without making the child available for hit testing, and without taking any
-/// room in the parent.
+/// Lays the child out as if it was in the tree, but without making the child
+/// available for hit testing, semantics and visiting via
+/// [Element.visitOnstageChildren].
 class RenderOffstage extends RenderProxyBox {
   /// Creates an offstage render object.
   RenderOffstage({
     bool offstage: true,
+    bool painted: false,
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: let's not use Offstage for something that's still visible. That's just weird.

**Comment by Hixie** [comment]
I don't think we should use the Offstage widget for this. It confuses the meaning of the term even further. Especially if you can be offstage and painted at the same time. What does that even mean.

**Comment by Hixie** [comment]
I think the way to do this is to have StatefulElement's visitOnstageChildren include its child only if an new getter State.onstage returns true.

Also, I know this API is technically available in release builds too for testing purposes, but I would still prefix it with "debug". This isn't something we want production code to be using, it really is just for testing.

**Comment by yjbanov** [comment]
> I don't think we should use the Offstage widget for this. It confuses the meaning of the term even further. Especially if you can be offstage and painted at the same time. What does that even mean.

The meaning I was going for is that the widget is painted for purposes other than interaction, such as an animated special effect, or caching. For example, a widget may not be interesting when obscured by another widget (the dropdown use-case), or mid-flight through a hero animation, or during screen transitions (e.g. tabs, navigation), or while fading out.

Let me send you two more versions, then we can compare.

**Comment by Hixie** [comment]
@yjbanov What is your plan regarding this PR?

**Comment by yjbanov** [comment]
Will close this and https://github.com/flutter/flutter/pull/11766 for now. https://github.com/flutter/flutter/pull/11767 solves the immediate issue with the dropdown, even if it may not offer a complete answer to the question "which elements are considered interactive?"

**Comment by Hixie** [comment]
Roger. We should probably keep the renaming though (`visitChildrenForSemantics` becoming `debugVisitOnstageChildren`)

**Comment by yjbanov** [comment]
Sounds good. I'll do the rename.

**Comment by yjbanov** [comment]
Only renaming and doc updates remain. PTAL.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 3]
```diff
@@ -2595,10 +2595,29 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
...
+  /// test framework to tap on a widget, by default the finder will look for
+  /// onstage elements and ignore the offstage ones.
+  ///
+  /// The default implementation defers to [visitChildren] and therefore treats
+  /// the element as onstage.
+  ///
+  /// See also:
+  ///
+  /// - [Offstage] widget that hides its children.
+  /// - [Finder][finder] that skips offstage widgets by default.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: `[Finder][finder]`?

File: packages/flutter/lib/src/widgets/framework.dart [hunk 4]
```diff
@@ -2595,10 +2595,29 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
...
+  /// - [RenderObject.visitChildrenForSemantics], in contrast to this method,
+  ///   designed specifically for excluding parts of the UI from the semantics
+  ///   tree.
+  ///
+  /// [finder]: https://docs.flutter.io/flutter/flutter_test/Finder-class.html
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: It should just work if you say `[Finder]`.

**Comment by Hixie** [comment]
LGTM

File: packages/flutter/lib/src/widgets/framework.dart (hunk 3 above)
Comment (yjbanov) [reply packages/flutter/lib/src/widgets/framework.dart]: Done

File: packages/flutter/lib/src/widgets/framework.dart (hunk 4 above)
Comment (yjbanov) [reply packages/flutter/lib/src/widgets/framework.dart]: Done

**Comment by yjbanov** [comment]
Heads up. I'm submitting this PR with incomplete Travis results after several unsuccessful attempts to restart the jobs. All jobs succeeded, except "analyze" and "tests" jobs on Mac that refuse to even begin. These same jobs passed on Linux.

**Pull request merged**